lib/libmp/src/core/mp_value.c

```diff
@@ -0,0 +1,773 @@
+/*
+ * Copyright 2025 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+
+#include "mp_value.h"
+#include "utils/mp_utils.h"
+
+#define DEFINE_COMPARE_SINGLE_VALUE_FUNC(typename, type)                                           \
+	static int mp_value_compare_##typename(const MpValue *value1, const MpValue *value2)       \
+	{                                                                                          \
+		type val1 = ((const MpValueSimple *)value1)->data.v_##typename;                    \
+		type val2 = ((const MpValueSimple *)value2)->data.v_##typename;                    \
+		if (val1 > val2) {                                                                 \
+			return MP_VALUE_GREATER_THAN;                                              \
+		}                                                                                  \
+                                                                                                   \
+		if (val1 < val2) {                                                                 \
+			return MP_VALUE_LESS_THAN;                                                 \
+		}                                                                                  \
+                                                                                                   \
+		return MP_VALUE_EQUAL;                                                             \
+	}
+
+typedef struct {
+	MpValue base;
+	union {
+		bool v_boolean;
+		int32_t v_int;
+		uint32_t v_uint;
+		const char *v_cstring;
+		MpObject *v_obj;
+		void *v_ptr;
+	} data;
+} MpValueSimple;
+
+typedef struct {
+	MpValue base;
+	sys_slist_t v_list;
+} MpValueList;
+
+typedef struct {
+	MpValue base;
+	union {
+		int32_t v_int;
+		uint32_t v_uint;
+	} min, max, step;
+} MpValueRange;
+
+typedef struct {
+	MpValue base;
+	union {
+		int32_t v_int;
+		uint32_t v_uint;
+	} num, denom;
+} MpValueFraction;
+
+typedef struct {
+	MpValue base;
+	MpValueFraction min, max, step;
+} MpValueFractionRange;
+
+typedef struct {
+	MpValue *value;
+	sys_snode_t node;
+} MpValueNode;
```

Comment:
It seems like all of these fits in almost a single 4 byte budget (except fractions).
This would make a [mem slab](https://docs.zephyrproject.org/latest/kernel/memory_management/slabs.html) allocator extremely fast with no fragmentation (and maybe smaller than a heap as no header structure required).